src/coreclr/jit/fgopt.cpp

```diff
@@ -5106,6 +5123,236 @@ void Compiler::fgMoveColdBlocks()
+            //
+            startBlock = block;
+            break;
+        }
+    }
+
+    if (startBlock == nullptr)
+    {
+        JITDUMP("\nSkipping reordering");
+        return;
+    }
+
+    // blockVector will contain the set of interesting blocks to move.
+    // tempBlockVector will assist with moving segments of interesting blocks.
+    //
+    BasicBlock**                blockVector     = new BasicBlock*[fgBBNumMax];
+    BasicBlock**                tempBlockVector = new BasicBlock*[fgBBNumMax];
+    unsigned                    blockCount      = 0;
+    ArrayStack<CallFinallyPair> callFinallyPairs(getAllocator(), hasEH ? ArrayStack<CallFinallyPair>::builtinSize : 0);
+
+    for (BasicBlock* const block : Blocks(startBlock, fgLastBBInMainFunction()))
+    {
+        // Don't consider blocks in EH regions
+        //
+        if (block->hasTryIndex() || block->hasHndIndex())
+        {
+            continue;
+        }
+
+        // We've reached the cold section of the main method body;
+        // nothing is interesting at this point
+        //
+        if (block->isRunRarely())
+        {
+            break;
+        }
+
+        blockVector[blockCount]       = block;
+        tempBlockVector[blockCount++] = block;
+
+        if (hasEH && block->isBBCallFinallyPair())
+        {
+            callFinallyPairs.Emplace(block, block->Next());
+        }
+    }
+
+    if (blockCount < 3)
+    {
+        JITDUMP("\nNot enough interesting blocks; skipping reordering");
+        return;
+    }
+
+    JITDUMP("\nInteresting blocks: [" FMT_BB "-" FMT_BB "]", startBlock->bbNum, blockVector[blockCount - 1]->bbNum);
+
+    auto evaluateCost = [](BasicBlock* const block, BasicBlock* const next) -> weight_t {
+        assert(block != nullptr);
+
+        if ((block->NumSucc() == 0) || (next == nullptr))
+        {
+            return 0.0;
+        }
+
+        const weight_t cost = block->bbWeight;
+
+        for (FlowEdge* const edge : block->SuccEdges())
+        {
+            if (edge->getDestinationBlock() == next)
+            {
+                return cost - edge->getLikelyWeight();
+            }
+        }
+
+        return cost;
+    };
+
+    // finalBlock is the first block after the set of interesting blocks.
+    // We will need to keep track of it to compute the cost of creating/breaking fallthrough into it.
+    // finalBlock can be null.
+    //
+    BasicBlock* const  finalBlock     = blockVector[blockCount - 1]->Next();
+    bool               improvedLayout = true;
+    constexpr unsigned maxIter        = 5; // TODO: Reconsider?
+
+    for (unsigned numIter = 0; improvedLayout && (numIter < maxIter); numIter++)
+    {
+        JITDUMP("\n\n--Iteration %d--", (numIter + 1));
+        improvedLayout              = false;
+        BasicBlock* const exitBlock = blockVector[blockCount - 1];
+
+        for (unsigned i = 1; i < (blockCount - 1); i++)
```

Comment:
I think the root of the TP cost is here -- we want to avoid having to search for possible cut points.

One approach is to just pick randomly, but I think we can do better for now. Roughly speaking in the pass above we should find all blocks that are either not just before their optimal successor and/or not just after their optimal successor.

We can rank these by the difference in the current vs optimal score. Then greedily pick the worst, that gives the first cut point. For the second cut point you can pick the best pred for the first cut point's current next block, or the best succ for the current pred of the first cut point's ideal successor. That is, if we have

```
S ~~~~ 1|2 ~~~ 3|4 ~~~ 5|6 ~~~ E

1's ideal succ is 4

reordering is

S ~~~~ 1|4 ~~~ 5|2 ~~~ 3|6 ~~~ E

So we either try and find a 5 which is the ideal pred of 2, or a 6 which is the ideal succ of 3.

Failing that we might pick some other block that is not currently followed by its ideal succ.
```
So one idea is to keep 3 values for each block: its min score, current score, and best score (lower is better). Order the blocks by current-min. Pick of the best as the first split, and then see if any of the next few provide a good second split.

Likely though this ends up needing a priority queue or similar as once we accept an arrangement we need to update some of the costings...